### PR TITLE
update napari to 0.1.3

### LINKED ISSUE #998: Is there a function that returns the outline of a segmented cell?
For example, after using Segmentation.Watershed to segment a cell, is there some function that takes the resulting np.ndarray as an input and returns another np.ndarray with the outlines of all segmented cells. This doesn't have to be the exact method; anything similar will do. 

I have tried looking around already for something similar, but I haven't been able to find it. If that function doesn't exist yet, I would like to help to write one since I need that function anyway.

Thank you!

Edit: To give some background to this, the reason I want to do this is because I want to detect and count dots that are only contained within cells, so I would use the outlines of cells from Segmentation.Watershed to crop the dots image to only include those outlined areas, and then count dots in those parts of the image. It is entirely possible that there is a function that does that already that I have just not been able to find yet. If that exists yet, please let me know.

### LINKED ISSUE #1422: Visualize Segmentation mask with original Images with starfish.display
Addresses needs in #1414

### PR DESCRIPTION
**Pull request by kne42**
- update napari version to `0.1.3`
- add segmentation masks visualization support (resolves #1422)

closes #998

### CODE REVIEW

**Comment by kne42** [comment]
when reviewing, please check that the axis ordering is right. not sure if I did it correctly

(example with ISS pipeline)
Screen Shot 2019-06-28 at 3 02 11 PM

File: starfish/core/_display.py [hunk 1]
```diff
@@ -190,49 +194,52 @@ def display(
...
-    if Window is None or Viewer is None:
-        warnings.warn("Requires napari 0.0.6. Run `pip install starfish[napari]` to install the "
-                      "necessary requirements.")
-        return
-
-    from PyQt5.QtWidgets import QApplication
+    try:
+        import napari
+        if napari.__version__ != NAPARI_VERSION:
+            raise Exception
```
Comment (ttung) [inline starfish/core/_display.py]: I would prefer that this be a specific exception type, or an exception with a specific payload that can be inspected to insure that it's what we expect.

File: starfish/core/_display.py [hunk 2]
```diff
@@ -190,49 +194,52 @@ def display(
...
+    except Exception:
```
Comment (ttung) [inline starfish/core/_display.py]: This is overly broad.  If there are other issues (perhaps `napari.__version__` doesn't exist?), it is improperly caught.

File: starfish/core/segmentation_mask.py [hunk 3]
```diff
@@ -154,6 +170,38 @@ def from_label_image(
 
         return cls(masks)
 
+    def to_label_image(
+            self,
+            shape: Optional[Tuple[int]] = None,
```
Comment (ttung) [inline starfish/core/segmentation_mask.py]:
```suggestion
            shape: Optional[Tuple[int, ...]] = None,
```

Comment (ttung) [reply starfish/core/segmentation_mask.py]: Would it be accurate to describe `shape` as kind of a crop operation?

File: starfish/core/segmentation_mask.py [hunk 4]
```diff
@@ -154,6 +170,38 @@ def from_label_image(
...
+            *,
+            ordering: List[Axes] = [Axes.ZPLANE, Axes.Y, Axes.X]
```
Comment (ttung) [inline starfish/core/segmentation_mask.py]:
```suggestion
            ordering: Sequence[Axes] = (Axes.ZPLANE, Axes.Y, Axes.X)
```

rationale: default parameter values should be immutable, as they are only created once. if the method modifies `ordering`, then on the subsequent run, `ordering` will have the wrong default value.

```
In [1]: def method(default_param=[1, 2, 3]):
   ...:     default_param.append(4)
   ...:     print(default_param)
   ...:     

In [2]: method()
[1, 2, 3, 4]

In [3]: method()
[1, 2, 3, 4, 4]

In [4]: 
```

File: starfish/core/segmentation_mask.py [hunk 5]
```diff
@@ -101,6 +111,12 @@ def append(self, mask: xr.DataArray):
         validate_segmentation_mask(mask)
         self._masks.append(mask)
 
+        for axis in Axes:
+            if axis.value in mask.coords:
+                max_val = mask.coords[axis.value].values[-1]
+                if self.max_shape[axis] is None or max_val > self.max_shape[axis]:
```
Comment (ttung) [inline starfish/core/segmentation_mask.py]:
```suggestion
                if self.max_shape[axis] is None or max_val >= self.max_shape[axis]:
```

if the max_val is the same as max_shape, i think you need to up max_shape.

File: starfish/core/_display.py [hunk 6]
```diff
@@ -258,13 +265,21 @@ def display(
...
+            coords=coords,
+            face_color="red",
+            edge_color="red",
+            symbol="ring",
+            size=sizes * radius_multiplier,
+            n_dimensional=True,
+            name='spots'
+        )
+
+    if masks is not None:
```
Comment (shanaxel42) [inline starfish/core/_display.py]: could be simplified to just `if masks`

File: starfish/core/_display.py (hunk 2 above)
Comment (kne42) [reply starfish/core/_display.py]: This is intentional since we pinned napari to a specific version. `napari.__version__` not existing or some other unforeseen error occurring would mean that it is not behaving as expected and must not be the pinned version. Do you still think this is bad?

File: starfish/core/segmentation_mask.py (hunk 3 above)
Comment (kne42) [reply starfish/core/segmentation_mask.py]: Ah, that's not what I was thinking of it as, and it would probably error if so. I was thinking more of when someone wanted to superimpose it on the original image so the shapes would match. Is this useful functionality?

File: starfish/core/segmentation_mask.py (hunk 4 above)
Comment (kne42) [reply starfish/core/segmentation_mask.py]: this method doesn't mutate the default argument but I suppose it's better practice so I'll do it

File: starfish/core/segmentation_mask.py (hunk 5 above)
Comment (kne42) [reply starfish/core/segmentation_mask.py]: wait could you explain this more? I don't quite understand

**Comment by kne42** [comment]
Thanks for the reviews! Will be addressing soon

File: starfish/core/segmentation_mask.py [hunk 7]
```diff
@@ -101,6 +111,12 @@ def append(self, mask: xr.DataArray):
         validate_segmentation_mask(mask)
         self._masks.append(mask)
 
+        for axis in Axes:
+            if axis.value in mask.coords:
+                max_val = mask.coords[axis.value].values[-1]
```
Comment (ambrosejcarr) [inline starfish/core/segmentation_mask.py]: why not the last coords value?

File: starfish/core/_display.py (hunk 2 above)
Comment (ttung) [reply starfish/core/_display.py]: I always prefer to catch the most specific exception possible because I don't want to mask some other error that may warrant a different fix.  In this case, your point makes a lot of sense for end users.

Two places this breaks down:
1. A developer who is experimenting with the starfish and napari bridge may find this problematic however.  Imagine that I have napari installed from a working tree, and I'm mucking around with the code there.  I could introduce a python parse error that this completely masks.
2. Someone introduces more code into this try-block, and doesn't think through the ramifications of having a catch-all exception handler around it.

tl;dr, I'm ok with you leaving this as is, because it's not a terribly common scenario that someone would encounter.  Be aware of doing general catch-all exception handlers though.

File: starfish/core/segmentation_mask.py (hunk 5 above)
Comment (ttung) [reply starfish/core/segmentation_mask.py]: So let's say max_vals are (15, 16, 13, 10).  The max shape, in this case, ought to be 17.

However, you would set max_shape to 16 when evaluating max_val=15, and then when you get to max_val=16, you won't bump it to 17.

File: starfish/core/segmentation_mask.py (hunk 3 above)
Comment (ttung) [reply starfish/core/segmentation_mask.py]: ohhh, i get it.  this _is_ useful functionality.

File: starfish/core/_display.py (hunk 2 above)
Comment (kne42) [reply starfish/core/_display.py]: What do you think about a catch-all that prepends the raised error message prior to the spiel about an incorrect napari version?

Comment (ttung) [reply starfish/core/_display.py]: If I were building this, I would do, (in pseudocode):

```
try:
    import napari
except ImportError:
    raise ImportError(f"Requires napari {NAPARI_VERSION...")

try:
    version = napari.__version__
except Exception as ex:
    raise Exception("Could not identify napari version") from ex

if version != NAPARI_VERSION:
        raise Exception(f"Incorrect version of napari installed ({version}).  Please run ...")
```

I think your proposed solution works too, or you could just chain the exception:

```
except Exception as ex:
    raise ImportError(f"...") from ex
```

File: starfish/core/_display.py [hunk 8]
```diff
@@ -258,13 +265,21 @@ def display(
...
 
-        viewer.add_markers(coords=coords, face_color="red", edge_color="red", symbol="ring",
-                           size=sizes * radius_multiplier, n_dimensional=True)
-
-    if new_viewer:
-        window.show()
-
-        if not INTERACTIVE:
-            app.exec()  # create blocking process to persist windows
+        viewer.add_markers(
```
Comment (sofroniewn) [inline starfish/core/_display.py]: now `add_points`

File: starfish/core/_display.py [hunk 9]
```diff
@@ -190,49 +194,52 @@ def display(
...
+            Axes.ZPLANE.value,
+            Axes.Y.value,
+            Axes.X.value
         ).values
 
-        layer = viewer.add_image(np.squeeze(reordered_array), multichannel=False)
+        layer = viewer.add_image(np.squeeze(reordered_array),
+                                 multichannel=False,
+                                 name='stack')
         layer.colormap = "gray"
```
Comment (sofroniewn) [inline starfish/core/_display.py]: now colormap default is gray, but can also pass `colormap` keyword arg to the above add_image function

File: starfish/core/_display.py (hunk 8 above)
Comment (sofroniewn) [reply starfish/core/_display.py]: also there is no longer a keyword arg `coords`, instead the first arg is the just data array, (can still be a variable named coords)

File: starfish/core/segmentation_mask.py (hunk 7 above)
Comment (kne42) [reply starfish/core/segmentation_mask.py]: you mean `mask.coords[axis.value][-1]`? it's because xarray is weird so a singular value returned is still an xarray, so we access it via the numpy array instead

**Comment by kne42** [comment]
while we wait on `0.1.2` to release I've updated to `0.1.1`. make sure that your vispy version is `0.5.3` before testing

the intensity table does not appear to be rendering correctly in the ISS notebook; am I doing something wrong? (cc @ambrosejcarr @sofroniewn)

**Comment by kne42** [comment]
@ambrosejcarr @ttung this is now ready for re-review and merge if everything looks okay to you :)

**Comment by ttung** [comment]
travis is still red fyi.

**Comment by dganguli** [comment]
It looks like Napari is handling overlapping cells in a funky way:

image

@sofroniewn seems to know what's up

**Comment by sofroniewn** [comment]
@kne42 so the issue in `to_label_image` is that with each new mask you are overwriting with the entire  area covered by the bounding box, whereas you want to just overwrite the area where the mask is not zero. Shouldn't be too hard to fix

File: starfish/core/segmentation_mask.py [hunk 10]
```diff
@@ -154,6 +170,43 @@ def from_label_image(
...
+            raise ValueError("shape less than the maximum of the data provided."
+                             "cropping is not supported at this time")
+
+        label_image = np.zeros(shape)
+
+        for i, mask in enumerate(self):
+            s = tuple(slice(mask.coords[o].values[0], mask.coords[o].values[-1] + 1)
+                      for o in ordering if o in mask.coords)
+            # align axes and insert into image
+            label_image[s] = (mask
```
Comment (sofroniewn) [inline starfish/core/segmentation_mask.py]: here need to just overwrite area where mask is True, not whole bounding box as defined in `s`

Comment (sofroniewn) [reply starfish/core/segmentation_mask.py]: Would this work

```python
        label_image = np.zeros(shape)

         for i, mask in enumerate(self):
            s = tuple(slice(mask.coords[o].values[0], mask.coords[o].values[-1] + 1)
                      for o in ordering if o in mask.coords)
            # align axes and insert into image
             temp_label_image = np.zeros(shape)

            temp_label_image[s] = (mask.transpose(*[o.value for o in ordering if o in mask.coords]).where(mask == False, i + 1))  # noqa: E712

label_image[temp_label_image > 0] = i + 1
```

Comment (kne42) [reply starfish/core/segmentation_mask.py]: that would work but it's memory inefficient

Comment (sofroniewn) [reply starfish/core/segmentation_mask.py]: sure - you can also convert everything to linear indices and then add the appropriate offsets based on the `mask.coords[o].values[0]` values

Comment (kne42) [reply starfish/core/segmentation_mask.py]: yeah, that's true

Comment (kne42) [reply starfish/core/segmentation_mask.py]: I've just updated, tell me if that fixes it

Comment (dganguli) [reply starfish/core/segmentation_mask.py]: it's fixed! thanks!
